File: src/subnetDistributor.ts
/*
 * Copyright 2018, James Nugent.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

// import * as aws from "@pulumi/aws";
import * as cidr from "./cidr";

/**
 * A SubnetDistributor is used to split a given CIDR block into a number of
 * subnets.
 */
export class SubnetDistributor {
    private readonly _privateSubnets: string[];
    private readonly _publicSubnets: string[];

    /**
     * Returns a subnet distributor configured to split the baseCidr into one
     * pair of public/private subnets for each availability zone in the
     * provider-configured region.
     * @param {string} baseCidr The CIDR block on which to base subnet CIDRs
     * @returns {Promise<SubnetDistributor>} A SubnetDistributor instance.
     */
    public static async perAz(baseCidr: string): Promise<SubnetDistributor> {
        const aws = await import("@pulumi/aws");
        const azCount = (await aws.getAvailabilityZones({
            state: "available",
        })).names.length;

        return new SubnetDistributor(baseCidr, azCount);
    }

    /**
     * Returns a subnet distributor configured to split the baseCidr into a fixed
     * number of public/private subnet pairs.
     * @param {string} baseCidr The CIDR block to split.
     * @param {number} azCount The number of subnet pairs to produce.
     * @returns {SubnetDistributor} A SubnetDistributor instance.
     */
    public static fixedCount(baseCidr: string, azCount: number): SubnetDistributor {
        return new SubnetDistributor(baseCidr, azCount);
    }

    /** @internal */
    private constructor(baseCidr: string, azCount: number) {
        const newBitsPerAZ = Math.log2(nextPow2(azCount));

        const azBases = [...Array(azCount).keys()].map((_, index) => {
            return cidr.subnetV4(baseCidr, newBitsPerAZ, index);
        });

        this._privateSubnets = azBases.map((block) => {
            return cidr.subnetV4(block, 1, 0);
        });

        this._publicSubnets = this._privateSubnets.map((block) => {
            const splitBase = cidr.subnetV4(block, 0, 1);
            return cidr.subnetV4(splitBase, 2, 0);
        });
    }

    /**
     * Returns an array of the CIDR blocks for the private subnets.
     * @returns {Promise<string[]>}
     */
    public async privateSubnets(): Promise<string[]> {
        return this._privateSubnets.slice();
    }

    /**
     * Returns an array of the CIDR blocks for the public subnets.
     * @returns {Promise<string[]>}
     */
    public async publicSubnets(): Promise<string[]> {
        return this._publicSubnets.slice();
    }
}

/**
 * nextPow2 returns the next integer greater or equal to n which is a power of 2.
 * @param {number} n input number
 * @returns {number} next power of 2 to n (>= n)
 */
function nextPow2(n: number): number {
    if (n === 0) {
        return 1;
    }

    n--;
    n |= n >> 1;
    n |= n >> 2;
    n |= n >> 4;
    n |= n >> 8;
    n |= n >> 16;

    return n + 1;
}

File: python/jen20_pulumi_aws_vpc/subnet_distributor.py
# Copyright 2018-2019, James Nugent.
#
# This Source Code Form is subject to the terms of the Mozilla Public License,
# v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain
# one at http://mozilla.org/MPL/2.0/.

"""
Contains utilities calculate appropriate CIDR address spaces from a base address
"""
import ipaddress
import math


class SubnetDistributor:
    """
    A SubnetDistributor divides a given CIDR block into `az_count` chunks - one
    per AWS availability zone - and then divides each chunk such that half of it
    is allocated to private addresses, one-quarter is allocated to public
    addresses, and the remaining quarter is left spare for future use.
    """

    @staticmethod
    def __next_power_of_2(number: int) -> int:
        return 1 << (number - 1).bit_length()

    @staticmethod
    def __cidr_subnet(base_address: str, prefix_extension: int, subnet_number: int) -> str:
        return str(list(ipaddress.ip_network(base_address).subnets(prefix_extension))[subnet_number])

    @staticmethod
    def __make_public_subnet(block: str) -> str:
        split_base = SubnetDistributor.__cidr_subnet(block, 1, 1)
        return SubnetDistributor.__cidr_subnet(split_base, 1, 0)

    @staticmethod
    def __make_private_subnet(block: str) -> str:
        return SubnetDistributor.__cidr_subnet(block, 1, 0)

    def __init__(self, base_cidr: str, az_count: int):
        new_bits_per_az = int(math.log(SubnetDistributor.__next_power_of_2(az_count), 2))
        az_bases = [SubnetDistributor.__cidr_subnet(base_cidr, new_bits_per_az, i) for i in range(az_count)]
        self.private_subnets = list([SubnetDistributor.__make_private_subnet(block) for block in az_bases])
        self.public_subnets = list([SubnetDistributor.__make_public_subnet(block) for block in az_bases])
